File: src/providers/tiptap/rules/addAttributes-missing-renderHTML.ts
/**
 * tiptap-addAttributes-missing-renderHTML (correctness)
 *
 * Detects TipTap node/mark attribute descriptors that define `parseHTML` but
 * not `renderHTML`. Without `renderHTML`, TipTap serializes the attribute
 * using the plain attribute name, while `parseHTML` reads from `data-*`.
 * The mismatch silently drops attribute values on HTML round-trips.
 */
import { findProperty, walk } from '../utils.js';

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'TipTap addAttributes descriptors with parseHTML must also define renderHTML',
      category: 'correctness',
      rationale:
        'When an attribute defines parseHTML: (el) => el.getAttribute("data-label") but no renderHTML, TipTap emits the attribute as a plain HTML attribute (label="...") instead of data-label="...". On re-parse, getAttribute("data-label") returns null and the value falls back to the default, silently discarding any customization across HTML export/import cycles.',
      docsUrl: 'https://tiptap.dev/docs/editor/extensions/custom-extensions/create-new/node#attributes',
      recommended: true,
    },
    messages: {
      missingRenderHTML:
        'Attribute defines parseHTML but no renderHTML. TipTap will serialize this attribute with the default name, breaking the HTML round-trip. Add renderHTML: (attrs) => ({ "data-<name>": attrs.<name> }).',
    },
    schema: [],
  },
  create(context: any) {
    function checkAddAttributesReturn(returnObj: any): void {
      if (returnObj?.type !== 'ObjectExpression') return;
      for (const prop of returnObj.properties ?? []) {
        if (prop?.type !== 'Property') continue;
        const val = prop.value;
        if (val?.type !== 'ObjectExpression') continue;
        // Check if descriptor has parseHTML but not renderHTML
        if (findProperty(val, 'parseHTML') && !findProperty(val, 'renderHTML')) {
          context.report({ node: val, messageId: 'missingRenderHTML' });
        }
      }
    }

    function checkAddAttributesFunction(fn: any): void {
      // Find ReturnStatement(s) and check the returned object
      walk(fn, (node: any) => {
        if (node?.type === 'ReturnStatement' && node.argument) {
          checkAddAttributesReturn(node.argument);
        }
      });
    }

    return {
      // Handles: addAttributes() { return { ... } } as method shorthand
      // and addAttributes: function() { } / addAttributes: () => { }
      Property(node: any) {
        const keyName =
          node.key?.type === 'Identifier'
            ? node.key.name
            : node.key?.type === 'Literal'
            ? node.key.value
            : null;
        if (keyName !== 'addAttributes') return;

        const val = node.value;
        if (!val) return;

        if (
          val.type === 'FunctionExpression' ||
          val.type === 'ArrowFunctionExpression'
        ) {
          // Arrow with expression body: addAttributes: () => ({ ... })
          if (val.body?.type === 'ObjectExpression') {
            checkAddAttributesReturn(val.body);
          } else {
            checkAddAttributesFunction(val.body);
          }
        }
      },
    };
  },
};

export const tiptapAddAttributesMissingRenderHTMLRule = rule;
export default rule;

File: src/providers/tiptap/rules/python/add_attributes_missing_render_html.py
"""tiptap-addAttributes-missing-renderHTML (python analog, correctness)

Tiptap's `addAttributes()` descriptors are plain JS objects, so a Python
backend that mirrors the same extension schema as data (e.g. a config file
validated/generated server-side, or a schema shared between a Python CMS and
the Tiptap frontend) uses the identical `parseHTML`/`renderHTML` vocabulary in
a dict literal.

A missing `renderHTML` is not itself the defect. Tiptap's
`getRenderedAttributes` falls back to `{ [name]: attrs[name] }`, so a
descriptor whose `parseHTML` reads the same-named attribute round-trips
correctly. This flags only the asymmetry that actually loses data: a
`parseHTML` reading a *different* name than the one Tiptap writes back, or
reading from somewhere an attribute cannot be written back to at all.

Flags:
    {"label": {"parseHTML": "el => el.getAttribute('data-label')"}}

Does NOT flag:
    {"href": {"parseHTML": "el => normalize(el.getAttribute('href'))"}}
    {"label": {"parseHTML": "...", "renderHTML": "..."}}
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

from utils import dict_get_any, dict_has_any, key_name, literal_str, loc  # noqa: E402

RULE_KEY = "tiptap-addAttributes-missing-renderHTML"

# Sources no default attribute serialization can write back, whatever the
# attribute is called. Bracketed so they can never equal a real name.
_STYLE_SOURCE = "<style>"
_CONTENT_SOURCE = "<content>"
_UNKNOWN_SOURCE = "<unknown>"

# The descriptor bodies are JS snippets held as Python strings, so the reads
# have to be recovered from the snippet text rather than from an AST.
_GET_ATTRIBUTE = re.compile(r"\b(?:get|has)Attribute\s*\(\s*['\"]([^'\"]+)['\"]")
_GET_ATTRIBUTE_COMPUTED = re.compile(r"\b(?:get|has)Attribute\s*\(\s*(?!['\"])")
_DATASET = re.compile(r"\.dataset\s*(?:\.([A-Za-z_$][\w$]*)|\[\s*['\"]([^'\"]+)['\"]\s*\])")
_STYLE = re.compile(r"\.style\s*(?:\.[A-Za-z_$]|\[|\.getPropertyValue\s*\()")
_CONTENT = re.compile(r"\.(?:textContent|innerText|innerHTML|outerHTML)\b")


def _dataset_key_to_attribute(key: str) -> str:
    return "data-" + re.sub(r"[A-Z]", lambda m: "-" + m.group(0).lower(), key)


def _read_sources(snippet: str) -> set[str]:
    """Every source name the parseHTML snippet reads, lowercased."""
    sources: set[str] = set()

    for name in _GET_ATTRIBUTE.findall(snippet):
        sources.add(name.lower())
    if _GET_ATTRIBUTE_COMPUTED.search(snippet):
        sources.add(_UNKNOWN_SOURCE)
    for dotted, bracketed in _DATASET.findall(snippet):
        sources.add(_dataset_key_to_attribute(dotted or bracketed).lower())
    if _STYLE.search(snippet):
        sources.add(_STYLE_SOURCE)
    if _CONTENT.search(snippet):
        sources.add(_CONTENT_SOURCE)

    return sources


def _iter_attribute_descriptors(tree: ast.AST):
    """Yields (name, descriptor dict, extension renderHTML) per container."""
    for node in ast.walk(tree):
        if not isinstance(node, ast.Dict):
            continue
        attributes = dict_get_any(node, "addAttributes")
        if not isinstance(attributes, ast.Dict):
            continue
        extension_render_html = literal_str(dict_get_any(node, "renderHTML")) or ""
        for key, value in zip(attributes.keys, attributes.values):
            name = key_name(key)
            if name is not None and isinstance(value, ast.Dict):
                yield name, value, extension_render_html


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for name, descriptor, extension_render_html in _iter_attribute_descriptors(tree):
        parse_html = dict_get_any(descriptor, "parseHTML")
        if parse_html is None or dict_has_any(descriptor, "renderHTML"):
            continue

        # `rendered: false` opts the attribute out of serialization entirely.
        rendered = dict_get_any(descriptor, "rendered")
        if isinstance(rendered, ast.Constant) and rendered.value is False:
            continue

        snippet = literal_str(parse_html)
        if snippet is None:
            continue

        sources = _read_sources(snippet)
        # Nothing read from the DOM, or a name resolvable only at runtime.
        if not sources or _UNKNOWN_SOURCE in sources:
            continue
        # Reading the attribute's own name is what the default write-back emits.
        if name.lower() in sources:
            continue

        # The extension's own renderHTML may write the attribute back by hand,
        # making the per-attribute renderHTML redundant rather than missing.
        if name in extension_render_html or any(s in extension_render_html for s in sources):
            continue

        line, col, end_line, end_col = loc(descriptor)
        out.append(
            {
                "file": path,
                "line": line,
                "column": col,
                "endLine": end_line,
                "endColumn": end_col,
                "ruleKey": RULE_KEY,
            }
        )

    return out
